### DISCUSSION TITLE
Add JWT token to the XHR requests

### DISCUSSION THREAD

**Discussion by medilies** [opcodesio/log-viewer#202] [Q&A] [answered]
In my case, I have the token stored in `local storage` by the main SPA. How can I instruct this package SPA to use that token?

It would be cool if I could inject a script with a function like `setTokenHeader` and the package SPA calls it automatically if it exists.

So, Is this mechanism supported? Or are there alternatives? Or will it be supported?

**Comment by arukompas** [comment, marked as answer]
Hey @medilies 

Why is that JWT token required in the first place?

Log Viewer uses cookie-based authentication and should already work out of the box. If it doesn't, you can try deleting `config/log-viewer.php` configuration file and re-publishing it.

**Reply by medilies** [reply]
Our Laravel backend is only used for `stateless` API calls.

We build our SPA separately and put its static files in the public folder. Then users need to authenticate once to get a JWT token, which is stored in the browser's local storage to be used when consulting guarded routes.

The issue isn't with the `config`, we have it published from the latest version, we only added our `JWT` middleware to it and we wish to resolve the authentication with this mechanism.

I think it would be cool if we could extend the package after installation by providing specific callbacks, or at least be able to publish the resource folder.

**Reply by arukompas** [reply]
@medilies interesting! Well, at the moment that's not supported because there's no way to change/override how the frontend of the package works. This is a type of app that I haven't considered yet, since most authentications are cookie-based.

I can't promise anything any time soon, but I'll keep this in mind and let you know if there's a solution ready :) Otherwise, you're also free to experiment and suggest a PR! :)

**Reply by medilies** [reply]
Thank you @arukompas for clarifying the situation regarding the package. I presented the situation to my front-end colleagues and encouraged them to contribute to extending its behavior. Meanwhile, I'll continue hacking around with middleware and stuff until I bypass the blockage in a clean way.

**Reply by arukompas** [reply]
@medilies  I'm cooking something up and will be ready next week. Could it be as simple as this? (There might be more logic of course, but as long as it's synchronous then it would work)
```js
window.LogViewer.headers['Authorization'] = 'Bearer ' + localStorage.getItem('jwt');
```

**Reply by medilies** [reply]
This seems good, but the local storage key needs to be configurable, with this approach the package remains a plug-and-play package.

In the opposite case, we will be able to duplicate the `jwt` under the `key` indicated by the package

**Reply by medilies** [reply]
By the way, our current work around is based on setting the JWT as a cookie. intercept the cookie by a middleware and inject its value as an AUTH header before reaching the JWT middleware

**Reply by arukompas** [reply]
Oh yeah the storage key is not an issue. You'll be able to write whatever JavaScript needed to fetch that key.

**Reply by medilies** [reply]
Hello @arukompas is there any news about the feature? We just need to estimate when we push the package to prod if wait for it. Thank you :)

**Reply by arukompas** [reply]
Hey @medilies , the feature is already out there, but I forgot to mention it to you here! Sorry 😅

Have a look at it here - https://log-viewer.opcodes.io/docs/2.x/configuration/access-to-log-viewer#configuring-how-the-front-end-makes-api-requests

All you need is publish the `index.blade.php` where you'll be able to set up the additional headers used by the XHR requests from the frontend. Hopefully that is enough to pull them from the localStorage :) This script is executed before the Vue app is loaded/mounted.

Hope this helps!

**Reply by medilies** [reply]
No worries XD, but was it tested? I tried it and I don't see an authorization header added.

```html
<body class="h-full px-3 lg:px-5 bg-gray-100 dark:bg-gray-900">
    <div id="log-viewer" class="flex h-full max-h-screen max-w-full">
        <router-view></router-view>
    </div>

    <!-- Global LogViewer Object -->
    <script>
        window.LogViewer = @json($logViewerScriptVariables);
        const token = localStorage.getItem('OUR_TOKEN_KEY');

        console.log(token); // logs successfully 👍
        // Add additional headers for LogViewer requests like so:
        window.LogViewer.headers['Authorization'] = 'Bearer ' + token;
    </script>

    <script src="{{ asset(mix('app.js', 'vendor/log-viewer')) }}"
        onerror="alert('app.js failed to load. Please refresh the page, re-publish Log Viewer assets, or fix routing for vendor assets.')">
    </script>
</body>
```

## Debug

### Request headers

```txt
GET /log-viewer/api/folders?direction=asc HTTP/1.1
Accept: application/json, text/plain, */*
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9,fr-FR;q=0.8,fr;q=0.7,ar-DZ;q=0.6,ar;q=0.5
Cache-Control: no-cache
Connection: keep-alive
Cookie: XSRF-TOKEN=...; OUR_APP_session=...
Host: 127.0.0.1:8000
Pragma: no-cache
Referer: http://127.0.0.1:8000/log-viewer
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36
X-CSRF-TOKEN: ...
X-Requested-With: XMLHttpRequest
X-XSRF-TOKEN: ...
sec-ch-ua: "Google Chrome";v="111", "Not(A:Brand";v="8", "Chromium";v="111"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "Windows"
```

### Browser console

![image](https://user-images.githubusercontent.com/35309918/226721547-eaced0af-2510-468f-ad61-ea2341335bd1.png)

**Reply by arukompas** [reply]
@medilies , yes, definitely tested :) 

I just tested again, both while running on 127.0.0.1:8000 (artisan serve) like yours, and also on HTTPS using Valet, and the authorization header gets sent:

<img width="1728" alt="Screenshot 2023-03-22 at 10 15 15" src="https://user-images.githubusercontent.com/8697942/226841039-443b9372-ca71-43ea-81a4-6b21f9d9e1f8.png">

Have you published the new assets after updating the package?
```bash
php artisan log-viewer:publish
```

**Reply by medilies** [reply]
Aaaah yes, I completely forgot to update the frontend assets 💯. It works perfectly now, thank you for the awesome work :D

**Reply by arukompas** [reply]
Really happy it works, got me scared there for a minute xD
